ouroboros-consensus-test/bench-storage/Bench/Ouroboros/Consensus/Storage/LedgerDB/HD.hs

```diff
@@ -0,0 +1,631 @@
+  Command generator: Model
+-------------------------------------------------------------------------------}
+
+data Model k v = Model {
+    diffs         :: DS.DiffSeq 'TS.UTxO k v     -- ^ The current diff sequence
+  , tip           :: Int                         -- ^ The tip of the diff
+                                                 --   sequence
+  , backingValues :: TT.TableValues 'TS.UTxO k v -- ^ Values that have been
+                                                 --   flushed to a backing store
+  , keyCounter    :: Int                         -- ^ A counter used to generate
+                                                 --   unique keys
+  , nrGenerated   :: Int                         -- ^ A counter for the number
+                                                 --   of commands generated
+                                                 --   up until now
+  }
+  deriving stock Generic
+  deriving anyclass NFData
+
+genInitialModel :: (Eq v, Arbitrary v) => GenConfig -> Gen (Model Int v)
+genInitialModel GenConfig{nrInitialValues} = do
+    kvs <- mapM genKeyValue [0 .. nrInitialValues - 1]
+    pure $ Model mempty 0 (TT.valuesFromList kvs) nrInitialValues 0
+
+genKeyValue :: Arbitrary v => k -> Gen (k, v)
+genKeyValue x = (x,) <$> arbitrary
+
+{-------------------------------------------------------------------------------
+  Command generator: Main generators
+-------------------------------------------------------------------------------}
+
+-- | A stateful generator for sequences of commands.
+--
+-- Note: We return the @'Model'@ because we might want to check invariants on
+-- it.
+generator :: (Eq v, Arbitrary v) => GenConfig -> Gen ([Cmd 'New Int v], Model Int v)
+generator conf = do
+  m <- genInitialModel conf
+  runStateT (genCmds conf) m
+
+-- | Like @'generator'@, but the order of operations is deterministic.
+generator' :: (Eq v, Arbitrary v) => GenConfig -> Gen ([Cmd 'New Int v], Model Int v)
+generator' conf = do
+  m <- genInitialModel conf
+  runStateT (genCmds' conf) m
+
+-- | Adapt a @GenConfig@ured generator to multiply the configured @nrCommands@
+-- parameter with a factor that depends on the QuickCheck size parameter.
+makeSized ::
+     (GenConfig -> Gen a)
+  -> GenConfig
+  -> Gen a
+makeSized fgen conf = do
+  n <- getSize
+  k <- choose (0, n)
+  let
+    p :: Float
+    p =
+      if n == 0 then
+        0
+      else
+        fromIntegral k / fromIntegral n
+    nr :: Float
+    nr = fromIntegral (nrCommands conf)
+
+  fgen (conf {nrCommands = round $ p * nr})
+
+type CmdGen v a = StateT (Model Int v) Gen a
+
+-- | Generate a list of commands in non-deterministic order.
+genCmds :: (Eq v, Arbitrary v) => GenConfig -> CmdGen v [Cmd 'New Int v]
+genCmds conf = mapM (const $ genCmd conf) [1 .. nrCommands conf]
+
+-- | Generate a list of commands in deterministic order.
+genCmds' :: forall v. (Eq v, Arbitrary v) => GenConfig -> CmdGen v [Cmd 'New Int v]
```

Comment:
This function should be documented better -- it is not clear how the deterministic order is achieved.

Furthermore, "deterministic order" might be confusing: In this case, the order of commands is deterministic, but the inputs generated for each command are random.